dev/commRefactoringNG - Virtual Printer Enhancements

kantlivelong opened the pull request:
#### What does this PR do and why is it necessary?
Adds PTY and Serial support to the Virtual Printer from CLI as well as various enhancements.

#### How was it tested? How can it be tested by the reviewer?
You know :slightly_smiling_face: 

#### Any background context you want to provide?

#### What are the relevant tickets if any?

#### Screenshots (if appropriate)

#### Further notes

kantlivelong commented:
> Automatic PR Validation failed

Hey, laser lips! Your mamma was a snowblower!

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -193,7 +201,7 @@ def _copy_to_incoming(self):
                             received = bytes(line)
 
-                            print(f"<<< {received.rstrip()}")
+                            print(f"<<< {received.rstrip().decode('utf-8')}")
```
`to_str`

kantlivelong replied:
I assume you mean `to_unicode`? `to_str` is marked as deprecated and suggests to use `to_bytes` which encodes.

foosel replied:
Er, yes... sorry, got confused there.

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -270,24 +309,102 @@ def tcp_command(host, port):
 
             commands.append(uds_command)
 
+        ##~~ Unix PTY
+
+        @click.command("pty")
+        @click.option(
+            "--path", "path", type=click.Path(), default="/tmp/octoprint-virtual_printer"
+        )
+        def pty_command(path):
+            if os.path.islink(path):
+                realpath = os.path.realpath(path)
+                if not os.path.exists(realpath):
+                    os.unlink(path)
+                else:
+                    print(
+                        f"{path} already exists as a valid symlink. Unable to continue."
+                    )
+                    sys.exit(1)
+            elif os.path.exists(path):
+                print(f"{path} already exists. Unable to continue.")
+                sys.exit(1)
+
+            click.echo(f"Creating Unix PTY on {path}...")
+
+            mfd, sfd = pty.openpty()
+            sfd_path = os.ttyname(sfd)
+            os.symlink(sfd_path, path)
+            fcntl.fcntl(
+                mfd, fcntl.F_SETFL, fcntl.fcntl(mfd, fcntl.F_GETFL) | os.O_NONBLOCK
+            )
+            attrs = termios.tcgetattr(mfd)
+            attrs[3] = attrs[3] & ~termios.ECHO
+            termios.tcsetattr(mfd, termios.TCSADRAIN, attrs)
+
+            virtual = PTYVirtualPrinterWrapper(mfd, settings, datafolder)
+            virtual.wait()
+
+        commands.append(pty_command)
```
Needs to go into a wrapper so we don't try to add this on unsupported platforms.

Something like this should work:

``` python
try:
    import fcntl
    import pty
    import termios

    # further code here
except ImportError:
    # unsupported on this platform
```

kantlivelong replied:
Done

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -232,16 +241,46 @@ def __init__(self, socket, *args, **kwargs):
+                if self._socket in ready[1]:
+                    self._socket.sendall(data)
+
+        class PTYVirtualPrinterWrapper(VirtualPrinterWrapper):
```
Is it really a dedicated `PTYVirtualPrinterWrapper` or is `FileVirtualPrinterWrapper` maybe the better term?

kantlivelong replied:
I was just trying to keep things the same but File is definitely a better term. No arguments against changing it so I've done so.

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -178,9 +185,10 @@ def _copy_to_incoming(self):
                 try:
                     while True:
-                        try:
+                        with self._mutex:
```
This forces half-duplex on *every* wrapper. I'm not sure that's the correct solution here.

kantlivelong replied:
You are most certainly right. No need for this. Removed

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -136,12 +136,18 @@ def register_transport_hook(self, *args, **kwargs):
         ]
 
     def cli_commands(self, cli_group, pass_octoprint_ctx, *args, **kwargs):
+        import fcntl
         import os
+        import pty
         import select
         import socket
+        import stat
+        import sys
+        import termios
         import threading
```
`fcntl`, `pty`, `termios` need to move into a posix only section or this will make the whole thing fail hard on windows systems

kantlivelong replied:
Fixed

foosel commented on src/octoprint/plugins/virtual_printer/__init__.py:
```diff
@@ -270,24 +309,102 @@ def tcp_command(host, port):
+            )
             def uds_command(path):
+                if os.path.exists(path):
+                    fd = os.open(path, os.O_PATH)
+                    mode = os.fstat(fd).st_mode
+                    os.close(fd)
```
``` python
with os.open(path, os.O_PATH) as fd:
    mode = os.fstat(fd).st_mode
```

kantlivelong replied:
:+1:

foosel reviewed (changes requested):
Got a couple points, see line comments.

What I'm unsure about is the changed socket timeout handling that you also included. Was there a specific reason why you changed the code so it never raises `socket.timeout`?

kantlivelong commented:
> What I'm unsure about is the changed socket timeout handling that you also included. Was there a specific reason why you changed the code so it never raises `socket.timeout`?

Actually by my logic I should have also removed `ConnectionAbortedError`. The reason I removed it (and maybe should have instead relocated it) was to keep `VirtualPrinterWrapper` generic. Does moving `socket.timeout` and `ConnectionAbortedError` to `SocketVirtualPrinterWrapper` make sense?

cp2004 commented:
> I'm just a bot 🤖 that does automatic checks, a human will intervene if I've made a mistake.

Humans are *trying* to intervene, but you're not letting them @github-actions

foosel commented:
Nope, I'm actually testing something here, sorry for the noise. Trying to make this thing stop complaining about branches like this one, but that needed a code change.
